projects/ADuCM3029_demo_cn0414/src/platform_drivers.c

```diff
@@ -0,0 +1,996 @@
+uint8_t uart_line_index = 0; /* CLI buffer index */
+uint8_t uart_cmd = 0; /* CLI command receive flag */
+uint8_t uart_flag = 0;
+uint8_t uart_ping_flag = 0;
+uint8_t uart_pong_flag = 0;
+volatile uint8_t ping, pong;
+volatile uint8_t *ping_ptr = &ping;
+volatile uint8_t *pong_ptr = &pong;
+static volatile int8_t uart_tx_flying = 0;
+uint8_t spi_init_flag = 0;
+
+/* Master SPI device handle */
+ADI_SPI_HANDLE h_spi_device;
+/* Device memory required for operate Master SPI device */
+uint8_t master_spi_device_mem[ADI_SPI_MEMORY_SIZE];
+
+/* Handle for Flash device */
+ADI_FEE_HANDLE h_fee_device;
+/* Memory for Flash device */
+uint8_t fee_device_mem[ADI_FEE_MEMORY_SIZE];
+/* Hardware report variable */
+uint32_t fee_hardware_error;
+/* Flash write buffer */
+uint32_t fee_data_buffer[FLASH_BUFFER_SIZE] __attribute__ ((aligned (4)));
+/* Pointer to flash page address */
+uint32_t *page_address = (uint32_t*)FLASH_PAGE_ADDR;
+/* Flash parameters */
+uint32_t nStartPage, nEndPage, nBlockNum;
+
+/* Memory for GPIO device */
+uint8_t mem_gpio_handler[ADI_GPIO_MEMORY_SIZE];
+uint8_t gpio_init_flag = 0;
+
+/* Master I2C device handle */
+ADI_I2C_HANDLE master_i2c_dev;
+/* Device memory required for operate Master I2C device */
+uint8_t master_i2c_device_memory[ADI_I2C_MEMORY_SIZE];
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * UART callback function.
+ *
+ * @param [in] cb_param - Callback parameter passed by the application.
+ * @param [in] event    - Trigger ID code.
+ * @param [in] arg      - Callback parameter passed by the driver.
+ *
+ * @return void
+ */
+static void uart_callback(void *cb_param, uint32_t event, void *arg)
+{
+	switch(event) {
+	case ADI_UART_EVENT_RX_BUFFER_PROCESSED:
+		if(arg == (void *)&ping) {
+			uart_ping_flag = 2;
+		}
+		if(arg == (void *)&pong) {
+			uart_pong_flag = 2;
+		}
+		break;
+	case ADI_UART_EVENT_NO_RX_BUFFER_EVENT:
+		if (uart_ping_flag == 0) {
+			adi_uart_SubmitRxBuffer((ADI_UART_HANDLE const)h_uart_device,
+					(void *)ping_ptr, 1, DMA_NOT_USE);
+			uart_ping_flag = 1;
+		}
+		if (uart_pong_flag == 0) {
+			adi_uart_SubmitRxBuffer((ADI_UART_HANDLE const)h_uart_device,
+					(void *)pong_ptr, 1, DMA_NOT_USE);
+			uart_pong_flag = 1;
+		}
+		break;
+	case ADI_UART_EVENT_TX_BUFFER_PROCESSED:
+		uart_tx_flying--;
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * Initialize the I2C communication peripheral.
+ *
+ * @param [out] desc      - The I2C descriptor.
+ * @param [in] init_param - The structure that contains the I2C parameters.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t i2c_init(i2c_desc **desc, const i2c_init_param *param)
+{
+	int32_t ret;
+	struct i2c_desc *dev;
+
+	/* Allocate memory for I2C device descriptor */
+	dev = malloc(sizeof(*dev));
```

Comment:
Wouldn't it be better to use `calloc() `instead of `malloc()`? The reason is that `calloc()` initializes the allocated memory to 0, which is safer and probably desired in most cases.
I know that most no-os projects use `malloc()`, but maybe we should start a new trend :)

Reply:
I generally agree with using calloc over malloc, but, in this situation, does it still make sense to use calloc if all the members of dev are written right away?